### PR TITLE
Modify file ownership in localstack image to support Openshift deployments

### PR DESCRIPTION
**Pull request by crungehottman**
The Openshift documentation [1] states that "[...] directories and files that are written to by processes in the image must be owned by the root group and be read/writable by that group. Files to be executed must also have group execute permissions."

The official localstack image does not satisfy these requirements. The following  permission error prevents running localstack in Openshift:
```
Error starting infrastructure: [Errno 13] Permission denied: '/opt/code/localstack/localstack/infra/dynamodb/log4j2.xml' Traceback (most recent call last):
  File "/opt/code/localstack/localstack/services/infra.py", line 366, in start_infra
    do_start_infra(asynchronous, apis, is_in_docker)
  File "/opt/code/localstack/localstack/services/infra.py", line 427, in do_start_infra
    prepare_installation()
  File "/opt/code/localstack/localstack/utils/analytics/profiler.py", line 156, in wrapped
    return f(*args, **kwargs)
  File "/opt/code/localstack/localstack/services/infra.py", line 403, in prepare_installation
    install.install_components(apis)
  File "/opt/code/localstack/localstack/services/install.py", line 246, in install_components
    parallelize(install_component, names)
  File "/opt/code/localstack/localstack/utils/common.py", line 1261, in parallelize
    result = pool.map(func, list)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 768, in get
    raise self._value
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/opt/code/localstack/localstack/services/install.py", line 242, in install_component
    installer()
  File "/opt/code/localstack/localstack/services/install.py", line 203, in install_dynamodb_local
    save_file(log4j2_file, log4j2_config)
  File "/opt/code/localstack/localstack/utils/common.py", line 804, in save_file
    with open(file, mode) as f:
PermissionError: [Errno 13] Permission denied: '/opt/code/localstack/localstack/infra/dynamodb/log4j2.xml'

```

Note the permissions on the files in the k8s pod, and the Openshift user ID:
```
/opt/code/localstack $ id
uid=1005450000(1005450000) gid=0(root) groups=1005450000
/opt/code/localstack $ id localstack
uid=1001(localstack) gid=1001(localstack) groups=1001(localstack)
/opt/code/localstack $ ls -l /opt/code/localstack/localstack/infra/dynamodb/
total 3728
-rw-r--r--    1 root     root       3787565 Nov  6  2020 DynamoDBLocal.jar
drwxr-xr-x    2 localsta localsta      4096 May 28  2020 DynamoDBLocal_lib
-rw-r--r--    1 localsta localsta      8644 May 28  2020 LICENSE.txt
-rw-r--r--    1 localsta localsta      1856 May 28  2020 README.txt
-rw-r--r--    1 localsta localsta	344 Nov  6  2020 log4j2.xml
drwxr-xr-x    2 localsta localsta      4096 May 28  2020 third_party_licenses
```

To avoid the file permission error mentioned above, it is necessary for Openshift users to 
maintain a separate localstack image with modified file permissions.  

For example, I can successfully deploy localstack to my Openshift namespace using the following Dockerfile:
```
FROM docker.io/localstack/localstack:latest

RUN chgrp -R 0 /opt/code/localstack/localstack && \
    chmod -R g=u /opt/code/localstack/localstack

ENTRYPOINT ["docker-entrypoint.sh"]
```
I'm requesting to make the same file permissions modifications (those which allow me to run localstack in Openshift) in the official localstack image.

[1] https://docs.openshift.com/container-platform/4.10/openshift_images/create-images.html

### CODE REVIEW

**Comment by crungehottman** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by crungehottman** [comment]
recheck

File: Dockerfile
Comment (alexrashed) [inline Dockerfile:256]: Unfortunately, this would cause all files in there to invalidate in the last layer of the Docker image, increasing the image size quite a bit (because the last layer would contain all files in /opt/code/localstack/ - again, but with different permissions). We should try to change the permissions right when adding the code (by specifying the permissions when running the [COPY](https://docs.docker.com/engine/reference/builder/#copy) and [ADD](https://docs.docker.com/engine/reference/builder/#add) commands).

**Review by alexrashed — CHANGES_REQUESTED** [review]
Hi @crungehottman!

Thanks for your contribution and sorry for the late response. We are currently in the process of migrating to a new directory structure (https://github.com/localstack/localstack/pull/6302). This will introduce additional directories being used, where the permissions will need to be adjusted as well. So we might need to wait for it to be merged and adjust your contribution to also handle these additional directories (/usr/lib/localstack, /etc/locasltack, /var/lib/localstack).
In addition, your change at the end of the Dockerfile would cause all files in there to invalidate in the last layer of the Docker image (see the comment).

I'm happy to assist in making these changes once the previously mentioned PR is merged.

File: Dockerfile
Comment (crungehottman) [reply Dockerfile:256]: Hello, @alexrashed!

I attempted to specify the permissions in the COPY and ADD commands with `--chmod=777`. The `RUN make init` command seems to populate most of the contents of `/usr/lib/localstack`, so specifying the permissions in the COPY and ADD commands alone does not seem to fix the file permissions issue. 

I added a couple of `ls -l` and a `chmod` statements around  the `RUN make init` command to help illustrate my problem (see steps 45 and 48 in the output below):

```
#45 [stage-5  8/14] ADD --chmod=777 .filesystem/usr/lib/localstack /usr/lib/localstack
#45 DONE 0.2s

#46 [stage-5  9/14] RUN ls -l /usr/lib/localstack
#46 0.372 total 14120
#46 0.372 drwxrwxrwx. 1 root root      140 Aug 10 23:09 elasticsearch
#46 0.372 -rwxrwxrwx. 1 root root 14456354 Aug  9 20:16 localstack-utils-fat.jar
#46 0.372 drwxrwxrwx. 1 root root     6074 Aug 10 23:08 stepfunctions
#46 DONE 0.4s

#47 [stage-5 10/14] ADD --chmod=777 localstack/ localstack/
#47 DONE 0.2s

#48 [stage-5 11/14] RUN ls -l /usr/lib/localstack && make init && ls -l /usr/lib/localstack && chmod -R 777 /usr/lib/localstack && ls -l /usr/lib/localstack
#48 0.387 total 14120
#48 0.387 drwxrwxrwx. 1 root root      140 Aug 10 23:09 elasticsearch
#48 0.387 -rwxrwxrwx. 1 root root 14456354 Aug  9 20:16 localstack-utils-fat.jar
#48 0.387 drwxrwxrwx. 1 root root     6074 Aug 10 23:08 stepfunctions
#48 0.388 . .venv/bin/activate; python -m localstack.services.install libs
#48 1.284 Initializing installation.
#48 1.303 INFO:__main__:Downloading and installing local DynamoDB server. This may take some time.
#48 1.303 INFO:__main__:Downloading and installing local Installing lambda runtime server. This may take some time.
#48 1.304 INFO:__main__:Downloading and installing local KMS server. This may take some time.
#48 1.509 INFO:__main__:downloading kinesis-mock binary from https://github.com/etspaceman/kinesis-mock/releases/download/0.2.5/kinesis-mock-linux-amd64-static
#48 5.378 INFO:localstack_ext.bootstrap.install:Installing IoT rule engine. This may take a while.
#48 16.82 Done.
#48 16.88 total 23972
#48 16.88 drwxr-xr-x. 1 root root       40 Aug 10 23:09 amazon-kinesis-client
#48 16.88 -rwxr-xr-x. 1 root root 10079412 Aug 10 23:09 aws-lambda-rie
#48 16.88 drwxr-xr-x. 1 root root      268 Aug 10 23:09 dynamodb
#48 16.88 drwxrwxrwx. 1 root root      140 Aug 10 23:09 elasticsearch
#48 16.88 drwxr-xr-x. 1 root root       62 Aug 10 23:09 kinesis-mock
#48 16.88 drwxr-xr-x. 1 root root       50 Aug 10 23:09 kms
#48 16.88 -rw-r--r--. 1 root root     1951 Aug 10 23:09 lambda.cfn-response.js
#48 16.88 -rwxrwxrwx. 1 root root 14456354 Aug  9 20:16 localstack-utils-fat.jar
#48 16.88 drwxr-xr-x. 1 root root       76 Aug 10 23:09 node_modules
#48 16.88 -rw-r--r--. 1 root root      889 Aug 10 23:09 package-lock.json
#48 16.88 drwxrwxrwx. 1 root root     6074 Aug 10 23:09 stepfunctions
#48 16.89 total 23972
#48 16.89 drwxrwxrwx. 1 root root       40 Aug 10 23:09 amazon-kinesis-client
#48 16.89 -rwxrwxrwx. 1 root root 10079412 Aug 10 23:09 aws-lambda-rie
#48 16.89 drwxrwxrwx. 1 root root      268 Aug 10 23:09 dynamodb
#48 16.89 drwxrwxrwx. 1 root root      140 Aug 10 23:09 elasticsearch
#48 16.89 drwxrwxrwx. 1 root root       62 Aug 10 23:09 kinesis-mock
#48 16.89 drwxrwxrwx. 1 root root       50 Aug 10 23:09 kms
#48 16.89 -rwxrwxrwx. 1 root root     1951 Aug 10 23:09 lambda.cfn-response.js
#48 16.89 -rwxrwxrwx. 1 root root 14456354 Aug  9 20:16 localstack-utils-fat.jar
#48 16.89 drwxrwxrwx. 1 root root       76 Aug 10 23:09 node_modules
#48 16.89 -rwxrwxrwx. 1 root root      889 Aug 10 23:09 package-lock.json
#48 16.89 drwxrwxrwx. 1 root root     6074 Aug 10 23:09 stepfunctions
#48 DONE 17.0s
```
(I built the image with `make docker-build` and `Docker version 20.10.17, build 100c701`)

Is it acceptable to append `&& chmod -R 777 /usr/lib/localstack` to the `RUN make init` command (https://github.com/localstack/localstack/blob/master/Dockerfile#L245)?

Comment (alexrashed) [reply Dockerfile:256]: I see, `make init` is executed as root, so all resources created in the step would have the same permission.
Adding the chmod to the `make init` stage seems fine to me. Can you verify that this resolves your issues when running LocalStack within OpenShift (with its default security constraints enabled?

Comment (crungehottman) [reply Dockerfile]: Hi @alexrashed, please let me know if I could/should make further improvements to my PR. Thanks!

Comment (alexrashed) [reply Dockerfile]: Sorry for the late response, and thanks for the continuous effort to push this forward.
Unfortunately, I haven't had a chance to take a deep look, but you are changing the permissions on a lot more stages than only for the `make init` stage (as discussed above).
This is quite a bullet. It could be dangerous (since it just sets 777 on basically everything) and is only a part of the solution (since LocalStack is not tested to be executed as a non-root user, which means that some tasks will fail).
It also seems that some of the might not be necessary (f.e. on `/usr/lib/localstack`).

However, most (or all) of these commands are necessary. But since there are so many of them, I think a more fine-grained approach would be better. This could be done by creating the `localstack` user earlier in the `Dockerfile` (but at least in the base stage) and setting the user and the root group for `ADD` and `COPY` commands. This would then respect the permissions right away. This would also be a more robust approach considering future additions / changes of the `Dockerfile` since there won't be an automated OpenShift testing stage in the near future.

I hope you can agree - at least to a certain degree - and I am happy to discuss this further. 
Do you think the user-based approach would be feasible?

Comment (crungehottman) [reply Dockerfile]: > you are changing the permissions on a lot more stages than only for the make init stage (as discussed above)

Ah, sorry for the confusion -- simply appending `&& chmod -R 777 /usr/lib/localstack` to the `make init` stage (as discussed above) also works perfectly fine. The image size is just as large as the image produced by my original commit, though. I included the additional file permissions changes because it somehow reduced the image size (I don't exactly understand how, but it did).

> This is quite a bullet. It could be dangerous (since it just sets 777 on basically everything)

 `&& chmod -R 775 /usr/lib/localstack`  works too, I just used 777 because there are other instances of "chmod 777" for these directories earlier in the Dockerfile.

> It also seems that some of the might not be necessary (f.e. on /usr/lib/localstack).

`/usr/lib/localstack` is the only one that is necessary. More specifically `/usr/lib/localstack/dynamodb`, but my team only uses dynamodb and s3, so I'm not sure if other OpenShift users encounter different errors when using other services. 

The permissions error we hit only references `usr/lib/localstack/dynamodb/log4j2.xml` (the group is missing write permissions):

```
2022-08-26T19:02:12.409 DEBUG --- [d-4 (worker)] localstack.utils.run       : Executing command: ['unzip', '-o', 'StepFunctionsLocal.jar', 'META-INF/MANIFEST.MF']
2022-08-26T19:02:12.503 DEBUG --- [d-4 (worker)] localstack.utils.run       : Executing command: ['zip', 'StepFunctionsLocal.jar', 'META-INF/MANIFEST.MF']
ERROR: '['zip', 'StepFunctionsLocal.jar', 'META-INF/MANIFEST.MF']': exit code 15; output: b'zip I/O error: Permission denied\nzip error: Could not create output file (StepFunctionsLocal.jar)\n'
Error starting infrastructure: [Errno 13] Permission denied: '/usr/lib/localstack/dynamodb/log4j2.xml' Traceback (most recent call last):
  File "/opt/code/localstack/localstack/services/infra.py", line 419, in start_infra
    thread = do_start_infra(asynchronous, apis, is_in_docker)
  File "/opt/code/localstack/localstack/services/infra.py", line 520, in do_start_infra
    prepare_installation()
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 77, in wrapped
    return f(*args, **kwargs)
  File "/opt/code/localstack/localstack/services/infra.py", line 474, in prepare_installation
    install.install_components(apis)
  File "/opt/code/localstack/localstack/services/install.py", line 750, in install_components
    parallelize(install_component, names)
  File "/opt/code/localstack/localstack/utils/threads.py", line 135, in parallelize
    return pool.map(func, arr)
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/opt/code/localstack/localstack/services/install.py", line 746, in install_component
    installer()
  File "/opt/code/localstack/localstack/services/install.py", line 565, in install_dynamodb_local
    save_file(log4j2_file, log4j2_config)
  File "/opt/code/localstack/localstack/utils/files.py", line 64, in save_file
    with open(file, mode, opener=_opener if permissions else None) as f:
PermissionError: [Errno 13] Permission denied: '/usr/lib/localstack/dynamodb/log4j2.xml'
```

> This could be done by creating the localstack user earlier in the Dockerfile (but at least in the base stage) and setting the user and the root group for ADD and COPY commands. This would then respect the permissions right away.

The file ownership changes appear to be fixed now actually, now we just need to give group write permissions on `/usr/lib/localstack`. I don't think the user approach alone would accomplish this; I think we'd still need a chmod.

I could certainly be missing something, but I'm assuming I'd alter the permissions like so: `ADD --chmod 777 --chown localstack:root .filesystem/usr/lib/localstack /usr/lib/localstack`. 

I've tried this approach:

```
Regarding #44 [stage-5  7/13] ADD --chown=localstack:root --chmod=777 .filesystem/usr/lib/localstack /usr/lib/localstack
#44 DONE 0.4s

#45 [stage-5  8/13] RUN ls -l /usr/lib/localstack
#45 0.385 total 14120
#45 0.385 drwxr-xr-x. 1 root       root      140 Aug 26 17:24 elasticsearch
#45 0.385 -rwxrwxrwx. 1 localstack root 14456354 Aug  9 20:16 localstack-utils-fat.jar
#45 0.385 drwxrwxrwx. 1 localstack root     6074 Aug 26 17:24 stepfunctions
#45 DONE 0.5s
```
Once the image is built, the earlier permissions and the file ownership changes do not apply (because of files later added in the `make init` stage).

```
root@e1d04db22b26:/opt/code/localstack# ls -l /usr/lib/ | grep localstack
drwxrwxrwx. 1 root root  318 Aug 26 17:27 localstack
root@e1d04db22b26:/opt/code/localstack# ls -l /usr/lib/localstack
total 23972
drwxr-xr-x. 1 root       root       40 Aug 26 17:27 amazon-kinesis-client
-rwxr-xr-x. 1 root       root 10079412 Aug 26 17:26 aws-lambda-rie
drwxr-xr-x. 1 root       root      268 Aug 26 17:26 dynamodb
drwxr-xr-x. 1 root       root      140 Aug 26 17:24 elasticsearch
drwxr-xr-x. 1 root       root       62 Aug 26 17:26 kinesis-mock
drwxr-xr-x. 1 root       root       50 Aug 26 17:26 kms
-rw-r--r--. 1 root       root     1951 Aug 26 17:26 lambda.cfn-response.js
-rwxrwxrwx. 1 localstack root 14456354 Aug  9 20:16 localstack-utils-fat.jar
drwxr-xr-x. 1 root       root       76 Aug 26 17:26 node_modules
-rw-r--r--. 1 root       root      889 Aug 26 17:26 package-lock.json
drwxrwxrwx. 1 localstack root     6074 Aug 26 17:27 stepfunctions
```
Because the `dynamodb` directory and files are added to `/usr/lib/localstack` by the `make init` command, and not by the `COPY` or `ADD` commands, I don't think altering the permissions with the `COPY` or `ADD` commands alone will suffice.

It seems that `chown` and `chmod` only apply to files that already exist in that directory, and do not apply to future files that will later be added to that directory. So when `make init` adds the files to `/usr/lib/localstack`, the previous `--chmod=775 --chown=localstack:root` does not apply. ACLs and umasks are the only two ways that I'm familiar with to set default file permissions on directories. Neither one appears to have any effect in Dockerfiles, though. A `chmod -R` near the seems to be the only way to add write permissions to `/usr/lib/localstack/dynamodb`, in my testing at least. 

Please do let me know if I'm missing something, thanks!

Comment (alexrashed) [reply Dockerfile]: I'm really sorry this PR takes so long, but there currently are quite a lot of moving parts in this area.
With https://github.com/localstack/localstack/pull/6755 we removed the execution of `make init` locally, such that it is only executed once - in a build step within the Docker build process.
In addition, there are efforts going on to ensure that there are no runtime failures in case a backend (like dynamodb local) is already installed but does not have permissions to be modified: https://github.com/localstack/localstack/pull/6785

With these new changes, I think it might work out of the box with the chmod on `make init` (I've seen you already rebased your changes and removed the other chmod commands). But I think now it would also work to change the user for the make init call.
This should look something like this:
```
USER localstack:root
RUN make init
USER root
```

Comment (crungehottman) [reply Dockerfile]: I am happy to report that  `localstack/localstack:1.1.0` works right out of the box on OpenShift -- no modifications required. Thank you for your help :).

Given that the reason I submitted this PR appears to have been resolved, feel free to close this PR (unless you think changing the user to `localstack:root` for the `make init` call might help to future-proof LocalStack deployments in OpenShift).

**Comment by alexrashed** [comment]
> I am happy to report that localstack/localstack:1.1.0 works right out of the box on OpenShift -- no modifications required. Thank you for your help :).
> Given that the reason I submitted this PR appears to have been resolved, feel free to close this PR (unless you think changing the user to localstack:root for the make init call might help to future-proof LocalStack deployments in OpenShift).

That is _great_ news!
Short recap of the thread above: The permission issues with LocalStack deployments in OpenShift have been resolved due to https://github.com/localstack/localstack/pull/6755 and https://github.com/localstack/localstack/pull/6785.
Thanks a lot for validating the fixes, since we don't have direct access to an OpenShift cluster. And since we don't have any automated tests in such a cluster, we will not change the permissions for now.
Please let us know if you experience any other issues.

**Pull request closed without merging**
